### DISCUSSION TITLE
running httpd on ports 80/443 *with* sudo?

### DISCUSSION THREAD

**Discussion by tjluoma** [Homebrew/discussions#1894] [Everyday usage]
I have a Mac server that has previously run using Apple's built-in Server software, especially the webserver.

However, since High Sierra, Apple no longer supports this.

Therefore, I would like to install and use Apache (`httpd`) via `brew` instead.

When I installed it, it printed this note:

```
The default ports have been set in /usr/local/etc/httpd/httpd.conf to 8080 and in
/usr/local/etc/httpd/extra/httpd-ssl.conf to 8443 so that httpd can run without sudo.

To restart httpd after an upgrade:
  brew services restart httpd
Or, if you don't want/need a background service you can just run:
  /usr/local/opt/httpd/bin/httpd -D FOREGROUND
```

It also created the file `~/Library/LaunchAgents/homebrew.mxcl.httpd.plist` which I assume is part of `brew services`.

To use this as a regular server, I need to be able to use port 80 and 443.

On macOS Server, `httpd` runs as the user `_www` not root, so I'm trying to figure out how to use port 80 and 443 without running `httpd` as root, and so far Google has not been any help (everything seems to be pointing in the other direction, for people who _don't_ want to use port 80/443).

Does anyone have any pointers on how to do this?

I've really enjoyed being able to use macOS as my webserver rather than dealing with switching to some sort of Linux hosting, and would like to keep doing so, but running an outdated/no longer supported OS seems like a bad idea for a server.

**Comment by SMillerDev** [comment]
I'm no httpd expert, but I think you can just change the config file to use those ports, and to assume the `_www` user. You will need to run the software as root for it to assume different users though.

**Comment by gromgit** [comment]
> `httpd` runs as the user `_www` not root, so I'm trying to figure out how to use port 80 and 443 without running `httpd` as root

Short answer: Forward ports 80 and 443 to other ports...though it's generally better to just use root.

Longer answer: Stuff like running daemons as a different user without requiring authentication, and opening ports below 1024, are _privileged operations_ that require root. Well-behaved daemons like (I assume) Apache HTTPD generally bunch all their privileged operations together, to be run as early as possible, then "drop privileges" by switching to a different user like `_www`.

For the specific issue of opening privileged ports, you can technically work around it by telling macOS's _packet filter_ to redirect traffic from specific ports to other ports. A worked example can be found [here](https://stackoverflow.com/a/35259958/4432299), but note that configuring this _itself_ requires root, so it's not quite a turnkey "zero-root" solution. You're also likely to forget about this hack in 6 months' time, and start wondering why you can't get anything to listen on port 80 any more. After all, if you're configuring httpd, why would you even _think_ of looking in `/etc/pf.conf`?

In contrast, configuring `httpd.conf` to listen on port 80, and `sudo brew services start httpd` to run it as root, is far more obvious and maintainable in the long term.

**TCIY** (The Choice Is Yours).
